push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/TokenLoader.java

```diff
@@ -169,4 +188,42 @@ public void loadAndQueueTokenBatch(@Observes @Dequeue MessageHolderWithVariants
             }
         }
     }
+
+    /**
+     * Tries to submit token batch to queue. Returns true in case of success.
+     *
+     * It retries submitting when it detects that queue is full until it times out and return false;.
+     *
+     * @return returns true if token batch was queued; return false is queue was still full even after retrying
+     */
+    private boolean tryToDispatchTokens(MessageHolderWithTokens msg) throws InterruptedException {
+        for (int i = 0 ; i <= MAX_TIME_TO_BLOCK_WHEN_QUEUE_IS_FULL; i += TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL) {
+            if (dispatchTokensOrFailWhenQueueIsFull(msg)) {
+                return true;
+            } else {
+                logger.fine(String.format("Queue is full, blocking loading of batch token #%s for %s variant (%s) for %s ms", msg.getSerialId(), msg.getVariant().getType().getTypeName(), msg.getVariant().getVariantID(), TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL));
+                Thread.sleep(TIME_BETWEEN_RETRIES_WHEN_QUEUE_IS_FULL);
```

Comment:
Not sure it is good idea, I was testing three options:

1) using `address-full-policy=BLOCK` blocks the thread until the queue has free space, which is resource wasting

2) using `address-full-policy=FAIL` will immediatelly fail MDB transaction, so the message will be redelivered (according to CLI settings, it will try it in 1 second, with exponential backoff 1.5x with each next failure)

I don't believe we need to fail immediatelly, since we may instead block for reasonable time and if the queue still hasn't space available, we will fail MDB transaction anyway. So last option I went with at the end is a hybrid:

3) block for one second and if queue is still full, we will fail (so we don't block too long, but we give it some time free space)

May be premature optimization though.